src/Data/Bin/Ordering.agda

```diff
@@ -0,0 +1,790 @@
+  where
+  open IneqReasoning
+
+------------------------------------------------------------------------------
+-- Monotonicity proofs for _*_, double.
+------------------------------------------------------------------------------
+
+*-mono-≤ :  _*_ Preserves₂ _≤_ ⟶ _≤_ ⟶ _≤_
+*-mono-≤ {x} {x'} {y} {y'} x≤x' y≤y' =  begin
+  x * y                  ≈⟨ sym $ cong₂ _*_ (fromℕ-toℕ x) (fromℕ-toℕ y) ⟩
+  fromℕ m * fromℕ n      ≈⟨ sym (fromℕ-homo-* m n) ⟩
+  fromℕ (m ℕ.* n)        ≤⟨ fromℕ-mono-≤ (ℕₚ.*-mono-≤ m≤m' n≤n') ⟩
+  fromℕ (m' ℕ.* n')      ≈⟨ fromℕ-homo-* m' n' ⟩
+  fromℕ m' * fromℕ n'    ≈⟨ cong₂ _*_ (fromℕ-toℕ x') (fromℕ-toℕ y') ⟩
+  x' * y'                ∎
+  where
+  open IneqReasoning
+  m    = toℕ x;              m'   = toℕ x'
+  n    = toℕ y;              n'   = toℕ y'
+  m≤m' = toℕ-mono-≤ x≤x';    n≤n' = toℕ-mono-≤ y≤y'
+
+*-monoʳ-≤ :  ∀ x → (x *_) Preserves _≤_ ⟶ _≤_
+*-monoʳ-≤ x {y} {y'} y≤y' =  *-mono-≤ {x} {x} {y} {y'} (≤-refl {x}) y≤y'
+
+*-monoˡ-≤ :  ∀ x → (_* x) Preserves _≤_ ⟶ _≤_
+*-monoˡ-≤ x {y} {y'} y≤y' =  *-mono-≤ {y} {y'} {x} {x} y≤y' (≤-refl {x})
+
+*-mono-< :  _*_ Preserves₂ _<_ ⟶ _<_ ⟶ _<_
+*-mono-< {x} {x'} {y} {y'} x<x' y<y' =  begin-strict
+  x * y                  ≈⟨ sym $ cong₂ _*_ (fromℕ-toℕ x) (fromℕ-toℕ y) ⟩
+  fromℕ m * fromℕ n      ≈⟨ sym (fromℕ-homo-* m n) ⟩
+  fromℕ (m ℕ.* n)        <⟨ fromℕ-mono-< (ℕₚ.*-mono-< m<m' n<n') ⟩
+  fromℕ (m' ℕ.* n')      ≈⟨ fromℕ-homo-* m' n' ⟩
+  fromℕ m' * fromℕ n'    ≈⟨ cong₂ _*_ (fromℕ-toℕ x') (fromℕ-toℕ y') ⟩
+  x' * y'                ∎
+  where
+  open IneqReasoning
+  m    = toℕ x;              m'   = toℕ x'
+  n    = toℕ y;              n'   = toℕ y'
+  m<m' = toℕ-mono-< x<x';    n<n' = toℕ-mono-< y<y'
+
+*-monoʳ-< :  ∀ x → ((suc x) *_) Preserves _<_ ⟶ _<_
+*-monoʳ-< x {y} {z} y<z =  begin-strict
+  (suc x) * y      ≈⟨ suc-* x y ⟩
+  y + xy           <⟨ +-mono-<-≤  {y} {z} {xy} {xz} y<z xy≤xz ⟩
+  z + xz           ≈⟨ sym (suc-* x z) ⟩
+  (suc x) * z      ∎
+  where
+  open IneqReasoning
+  xy = x * y;   xz = x * z;   xy≤xz = *-monoʳ-≤ x (<⇒≤ {y} {z} y<z)
+
+*-monoˡ-< :  ∀ x → (_* (suc x)) Preserves _<_ ⟶ _<_
+*-monoˡ-< x {y} {z} y<z = begin-strict
+  y * (suc x)   ≈⟨ *-comm y (suc x) ⟩
+  (suc x) * y   <⟨ *-monoʳ-< x {y} {z} y<z ⟩
+  (suc x) * z   ≈⟨ *-comm (suc x) z ⟩
+  z * (suc x)   ∎
+  where
+  open IneqReasoning
+
+double-mono-≤ :  double Preserves _≤_ ⟶ _≤_
+double-mono-≤ {x} {y} x≤y =  begin
+  double x   ≈⟨ double≗2B* x ⟩
+  2B * x     ≤⟨ *-monoʳ-≤ 2B {x} {y} x≤y ⟩
+  2B * y     ≈⟨ sym (double≗2B* y) ⟩
+  double y   ∎
+  where
+  open IneqReasoning
+
+double-mono-< :  double Preserves _<_ ⟶ _<_
+double-mono-< {x} {y} x<y =  begin-strict
+  double x   ≈⟨ double≗2B* x ⟩
+  2B * x     <⟨ *-monoʳ-< 1B {x} {y} x<y ⟩
+  2B * y     ≈⟨ sym (double≗2B* y) ⟩
+  double y   ∎
+  where
+  open IneqReasoning
+
+double-back-mono-≤ :  ∀ {x y} → double x ≤ double y → x ≤ y
+double-back-mono-≤ {x} {y} 2x≤2y  with <-cmp x y
+... | tri< x<y _   _   =  <⇒≤ {x} {y} x<y
```

Comment:
Are these implicits needed?

Reply:
Implicits removed from ``*-monoʳ-< 1B {x} {y} x<y`` and ``<⇒≤ {x} {y} x<y``.